lib/services/android-project-service.ts

```diff
@@ -407,23 +435,61 @@ export class AndroidProjectService extends projectServiceBaseLib.PlatformProject
 	}
 
 	public async beforePrepareAllPlugins(projectData: IProjectData, dependencies?: IDependencyData[]): Promise<void> {
+		const shouldUseNewRoutine = this.shouldUseNewRuntimeGradleRoutine(projectData);
+
 		if (dependencies) {
-			const platformDir = path.join(projectData.platformsDir, "android");
-			const buildDir = path.join(platformDir, "build-tools");
-			const checkV8dependants = path.join(buildDir, "check-v8-dependants.js");
-			if (this.$fs.exists(checkV8dependants)) {
-				const stringifiedDependencies = JSON.stringify(dependencies);
-				try {
-					await this.spawn('node', [checkV8dependants, stringifiedDependencies, projectData.platformsDir], { stdio: "inherit" });
-				} catch (e) {
-					this.$logger.info("Checking for dependants on v8 public API failed. This is likely caused because of cyclic production dependencies. Error code: " + e.code + "\nMore information: https://github.com/NativeScript/nativescript-cli/issues/2561");
+			dependencies = this.filterUniqueDependencies(dependencies);
+			if (shouldUseNewRoutine) {
+				this.provideDependenciesJson(projectData, dependencies);
+			} else {
+				// TODO: Remove prior to the 4.0 CLI release @Pip3r4o @PanayotCankov
+
+				const platformDir = path.join(projectData.platformsDir, AndroidProjectService.ANDROID_PLATFORM_NAME);
+				const buildDir = path.join(platformDir, "build-tools");
+				const checkV8dependants = path.join(buildDir, "check-v8-dependants.js");
+				if (this.$fs.exists(checkV8dependants)) {
+					const stringifiedDependencies = JSON.stringify(dependencies);
+					try {
+						await this.spawn('node', [checkV8dependants, stringifiedDependencies, projectData.platformsDir], { stdio: "inherit" });
+					} catch (e) {
+						this.$logger.info("Checking for dependants on v8 public API failed. This is likely caused because of cyclic production dependencies. Error code: " + e.code + "\nMore information: https://github.com/NativeScript/nativescript-cli/issues/2561");
+					}
 				}
 			}
 		}
 
-		const projectRoot = this.getPlatformData(projectData).projectRoot;
+		if (!shouldUseNewRoutine) {
+			// TODO: Remove prior to the 4.0 CLI release @Pip3r4o @PanayotCankov
+			const projectRoot = this.getPlatformData(projectData).projectRoot;
+			await this.cleanProject(projectRoot, projectData);
+		}
+	}
+
+	private filterUniqueDependencies(dependencies: IDependencyData[]): IDependencyData[] {
+		const depsDictionary = dependencies.reduce((dict, dep) => {
+			const collision = dict[dep.name];
+			// in case there are multiple dependencies to the same module, the one declared in the package.json takes precedence
+			if (!collision || collision.depth > dep.depth) {
+				dict[dep.name] = dep;
+			}
+			return dict;
+		}, <IDictionary<IDependencyData>>{});
+		return Object.keys(depsDictionary).map(key => depsDictionary[key]);
```

Comment:
this seems like you are trying to return array of the values in the dictionary. If that's the case, you can just use:
```TypeScript
_.values(depsDictionary)
```